### ISSUE TITLE
Cannot resolve module 'vertx'

### ISSUE THREAD

**Issue by lookapanda** [cujojs/when#482]
```
ERROR in ./~/when/lib/env.js
Module not found: Error: Cannot resolve module 'vertx' in /node/rubix/node_modules/when/lib
 @ ./~/when/lib/env.js 32:14-35
```

I always get this error. And I really don't know why exactly. All I did is just installed when with `npm install when`

**Comment by briancavalier** [comment]
@lookapanda That's a very odd error.  I'm not able to reproduce it (and surely we would have heard about it before since it appears to make when completely unusable!).  Could you provide a bit more info:

1. What version of node are you using?
2. What does this report:

```
$ node
> Object.prototype.toString.call(process)
```

Thanks!

**Comment by Mte90** [comment]
I get the same error and my system is debian sid.
```$  node
> Object.prototype.toString.call(process)
'[object process]'
```
The node version in my case is 5.7.1

**Comment by briancavalier** [comment]
Hmmm, is that error happening during application runtime?  Or is it happening at build time when using some sort of build tool, like webpack, browserify, rollup, etc?

If that error is being reported at runtime (as opposed to build time, e.g. browserify, webpack, etc.), the *only* way control could ever reach [this line](https://github.com/cujojs/when/blob/master/lib/env.js#L32) is if all three of these things are true:

1. The environment isn't recognized as Node, and
2. the environment doesn't have MutationObserver (obviously this is browser only), and
3. the environment doesn't have `setTimeout`

The only way to detect if code is running in vertx is to rule out other environments (using the tests above), and then assume vertx and just try to require() it.  However, build tools scan source code and ASTs looking for require statements in order to create bundles.

So, if that error is happening at build time, then I think the right thing to do is to configure your build tool to ignore `vertex`.

**Comment by Mte90** [comment]
I don't remember but i think that was a problem during the build time but I had an old version of npm and after the update the problem disappear.

**Comment by kriuz** [comment]
I'm seeing this error during build time with `node 4.2.2` and `npm 3.8.1`. I've gotten around this error by adding `{ vertx = 'commonjs vertx' }` to my webpack config file under `externals`.

**Comment by belenbarbed** [comment]
This was also found (and solved) in [Issue 305](https://github.com/stefanpenner/es6-promise/issues/305) of [stefanpenner](https://github.com/stefanpenner)'s [es6-promise package](https://github.com/stefanpenner/es6-promise) (patched in [this commit](https://github.com/stefanpenner/es6-promise/commit/90c0595230b2caa23eefdf9da3480322dbf4edf7)).

tl;dr he edited ```when/lib/env.js```:
```
} else if (!capturedSetTimeout) { // vert.x
-  var vertxRequire = require;
-  var vertx = vertxRequire('vertx');
+  var vertx = Function('return this')().require('vertx');
    setTimer = function (f, ms) { return vertx.setTimer(ms, f); };
    clearTimer = vertx.cancelTimer;
    asap = vertx.runOnLoop || vertx.runOnContext;
}
```

Understandably not the cleanest patch but it did the job.

**Comment by theMK2k** [comment]
I had the same issue and the fix proposed by @belenbarbed works. My setup is an electron application.

Any chance that this fix gets an implementation in the official `when` release?

**Comment by msyfls123** [comment]
@theMK2k I hardly expect that `when` would release a new version and make it works with `webpack`:
```javascript
webpack.config = {
  resolve: {
    alias: {
      vertx: 'path/to/stub-vertx',
    }
  }
};
```

```javascript
// stub-vertx.js
module.exports = this;
```

**Comment by trylaarsdam** [comment]
I've also solved this issue on my end by changing ./lib/env.js line 32:
from - `var vertx = vertxRequire('vertx');`
to - `var vertx =vertxRequire('@vertx/core');` and also installing `@vertx/core` with npm

Hopefully some patch will fix this at some point (however unlikely since this is nearly 5yrs old)
